Actions/Github-Helper.psm1

```diff
@@ -708,91 +708,240 @@ function Set-JsonContentLF {
+        }
+
+        $CICDRuns = @($workflowRuns.workflow_runs | Where-Object { $_.name -eq ' CI/CD' })
+
+        if ($CICDRuns.Count -gt 0) {
+            $lastSuccessfulCICDRun = $CICDRuns[0].id
+            Write-Host "Found last successful CICD run: $($LastSuccessfulCICDRun)"
+            break
+        }
+
+        $page += 1
+    }
+
+    return $lastSuccessfulCICDRun
+}
+
+<#
+    Gets the CICD run ID for the specified repository and branch, that built the specified project, mask and version.
+
+    If no successful CICD run is found, 0 is returned.
+#>
+function FindCICDRunForVersion {
+    Param(
+        [Parameter(Mandatory = $true)]
+        [string] $token,
+        [Parameter(Mandatory = $true)]
+        [string] $api_url,
+        [Parameter(Mandatory = $true)]
+        [string] $repository,
+        [Parameter(Mandatory = $true)]
+        [string] $branch,
+        [Parameter(Mandatory = $true)]
+        [string] $project,
+        [Parameter(Mandatory = $true)]
+        [string] $mask,
+        [Parameter(Mandatory = $true)]
         [string] $version
     )
 
     $headers = GetHeader -token $token
-    $total_count = 0
-    if ($version -eq 'latest') { $version = '*' }
-    # Download all artifacts matching branch and version
-    # We might have results from multiple workflow runs, but we will have all artifacts from the workflow run that created the first matching artifact
-    # Use the buildOutput artifact to determine the workflow run id (as that will always be there)
-    $artifactPattern = "*-$branch-*-$version"
-    # Use buildOutput artifact to determine the workflow run id to avoid excessive API calls
-    # Reason: A project called xx-main will match the artifact pattern *-main-*-version, and there might not be any artifacts matching the mask
-    $buildOutputPattern = "*-$branch-BuildOutput-$version"
-    # Old builds from PR runs are vresioned differently and should be ignored
-    $ignoreBuildOutputPattern1 = "*-$branch-BuildOutput-*.*.2147483647.0"
-    # Build Output from TestCurrent, TestNextMinor and TestNextMajor are named differently and should be ignored
-    $ignoreBuildOutputPattern2 = "*-$branch-BuildOutput-*-*"
-    Write-Host "Analyzing artifacts matching $artifactPattern"
+    $CICDRun = 0
+    $per_page = 100
+    $page = 1
+
+    # Construct the project artifact name
+    $artifactPattern = "$project-$branch-$mask-$version"
+    $escapedArtifactPattern = [Uri]::EscapeDataString($artifactPattern)
+
     while ($true) {
-        if ($total_count -eq 0) {
-            # First iteration - initialize variables
-            $matchingArtifacts = @()
-            $buildOutputArtifacts = @()
-            $per_page = 100
-            $page_no = 1
-        }
-        $uri = "$api_url/repos/$repository/actions/artifacts?per_page=$($per_page)&page=$($page_no)"
-        Write-Host $uri
+        $uri = "$api_url/repos/$repository/actions/artifacts?per_page=$($per_page)&page=$($page)&name=$escapedArtifactPattern"
```

Comment:
I don't think the name property support wildcards?
Is this the reason for specifying project[0]?`
Problem is here that the current version of the repo might have different projects than the workflow run you are looking for.
If you have added a project called ABC, you might be looking for project ABC version 1.2.3.4 - which doesn't exist, but project DEF and GHI does

Reply:
Yes, the name query parameter doesn't threat wild characters as such.

Using _BuildOutput_ would also work but build output artifacts are produced even on failed CICDs, so that's why I went with the first project in the list.

Are you saying that getting version _latest_ for project _ABC_ shouldn't look for the latest CICD run, but for the latest CICD that actually built project _ABC_? I don't think that should be supported now, should it? 🤔

Reply:
No, this comment is in FindCICDforVersion.
You call this function with projects[0] - which might be a * or a project name which doesn't exist.
This is the reason for not using the name parameter in the existing implementation.

Reply:
Oh, `$projects` can actually be a list of patterns itself? That I didn't even think of. 😄 

The project name is in every artifact name, so that approach won't work like this.
Back to introducing a <VERSION> artifact? Any other ideas?

Reply:
Reverted the usage of `FindLatestSuccessfulCICDRunForVersion`.

For now, latest successful CICD run is used only when version is "latest".